### PR TITLE
Optimize the cosine_similarity_top_k function performance

### PR DESCRIPTION
**Pull request by misrasaurabh1**
Optimizing important numerical code and making it run faster.

Performance went up by 1.48x (148%). Runtime went down from 138715us to 56020us

Optimization explanation:

The `cosine_similarity_top_k` function is where we made the most significant optimizations. 
Instead of sorting the entire score_array which needs considering all elements, `np.argpartition` is utilized to find the top_k largest scores indices, this operation has a time complexity of O(n), higher performance than sorting. Remember, `np.argpartition` doesn't guarantee the order of the values. So we need to use argsort() to get the indices that would sort our top-k values after partitioning, which is much more efficient because it only sorts the top-K elements, not the entire array. Then to get the row and column indices of sorted top_k scores in the original score array, we use `np.unravel_index`. This operation is more efficient and cleaner than a list comprehension. 

The code has been tested for correctness by running the following snippet on both the original function and the optimized function and averaged over 5 times.
```
def test_cosine_similarity_top_k_large_matrices():
    X = np.random.rand(1000, 1000)
    Y = np.random.rand(1000, 1000)
    top_k = 100
    score_threshold = 0.5
    gc.disable()
    counter = time.perf_counter_ns()
    return_value = cosine_similarity_top_k(X, Y, top_k, score_threshold)
    duration = time.perf_counter_ns() - counter
    gc.enable()
```

@hwaking @hwchase17 @jerwelborn 

Unit tests pass, I also generated more regression tests which all passed.

### CODE REVIEW

**Comment by misrasaurabh1** [comment]
The mypy isn't able to correctly detect the return type, the types should still be compatible. How should I fix the mypy warming?

File: libs/langchain/langchain/utils/math.py [hunk 1]
```diff
@@ -46,11 +46,11 @@ def cosine_similarity_top_k(
...
-    sorted_idxs = score_array.flatten().argsort()[::-1]
-    top_k = top_k or len(sorted_idxs)
-    top_idxs = sorted_idxs[:top_k]
     score_threshold = score_threshold or -1.0
-    top_idxs = top_idxs[score_array.flatten()[top_idxs] > score_threshold]
-    ret_idxs = [(x // score_array.shape[1], x % score_array.shape[1]) for x in top_idxs]
-    scores = score_array.flatten()[top_idxs].tolist()
-    return ret_idxs, scores
+    score_array[score_array < score_threshold] = 0
+    top_k = min(top_k or len(score_array), np.count_nonzero(score_array))
```
Comment (baskaryan) [inline libs/langchain/langchain/utils/math.py:51]: will np.count_nonzero be strictly less than len(score_array)? in which case latter isn't needed?

Comment (misrasaurabh1) [reply libs/langchain/langchain/utils/math.py:51]: The problem is that top_k can also be `None`, so the `or len(score_array)` is such that the other `np.count_nonzero(score_array)` always wins `min`
When `top_k` is not `None`, the `len` is just ignored.

**Comment by baskaryan** [comment]
thanks @misrasaurabh1!

**Pull request merged**